operator/pkg/ingress/secret.go

```diff
@@ -0,0 +1,363 @@
+}
+
+func (sm *syncSecretManager) handleSecretUpdatedEvent(ev secretUpdatedEvent) error {
+	if ev.oldSecret.GetNamespace() == sm.namespace || ev.newSecret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.syncSecret(ev.newSecret)
+}
+
+func (sm *syncSecretManager) handleSecretDeletedEvent(ev secretDeletedEvent) error {
+	if ev.secret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.deleteSecret(ev.secret)
+}
+
+func (sm *syncSecretManager) handleIngressAddedEvent(ev ingressAddedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.ingress)
+}
+
+func (sm *syncSecretManager) handleIngressUpdatedEvent(ev ingressUpdatedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.newIngress)
+}
+
+// handleIngressDeletedEvent is doing nothing right now as the underlying secrets could be
+// referenced in other Ingress resources, which is common use case for wildcard secrets (e.g
+// *.foo.com)
+func (sm *syncSecretManager) handleIngressDeletedEvent(ev ingressDeletedEvent) error {
+	return nil
+}
+
+func (sm *syncSecretManager) handleIngressUpsertedEvent(ingress *slim_networkingv1.Ingress) error {
+	if !tlsEnabled(ingress) {
+		return nil
+	}
+
+	for _, tls := range ingress.Spec.TLS {
+		// check if the secret is available
+		key := getSecretKey(ingress.GetNamespace(), tls.SecretName)
+		secret, exists, err := sm.getByKey(key)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("secert does not exist: %s", key)
+		}
+
+		sm.lock.Lock()
+		sm.watchedSecretMap[key] = getSyncedSecretKey(sm.namespace, secret.GetNamespace(), secret.GetName())
+		sm.lock.Unlock()
+
+		// proceed to sync secret
+		err = sm.syncSecret(secret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// syncSecret performs an upsert to make sure that secret in secret namespace is in synced with original secret
+func (sm *syncSecretManager) syncSecret(original *slim_corev1.Secret) error {
+	sm.lock.RLock()
+	defer sm.lock.RUnlock()
+
+	key := getSecretKey(original.GetNamespace(), original.GetName())
+	if _, exists := sm.watchedSecretMap[key]; !exists {
+		// the secret is not used in TLS, ignoring.
+		return nil
+	}
+	syncKey := getSyncedSecretKey(sm.namespace, original.GetNamespace(), original.GetName())
+
+	scopedLog := log.WithField("secret", key).WithField("synced-secret", syncKey)
+	syncedSecret, exists, err := sm.getByKey(syncKey)
+	if err != nil {
+		scopedLog.WithError(err).Errorf("unable to lookup secret")
+		return err
+	}
+
+	// create a new secret if not exits
+	if !exists {
+		newSecret := original.DeepCopy()
+		newSecret.Name = getSyncedSecretName(original.GetNamespace(), original.GetName())
+		newSecret.Namespace = sm.namespace
+
+		// create a new secret
+		_, err = sm.client.Create(context.TODO(), toV1Secret(newSecret), metav1.CreateOptions{})
+		if err != nil {
+			scopedLog.WithError(err).Errorf("unable to create secret")
+			return err
+		}
+		return nil
+	}
+
+	// update existing secret
```

Comment:
Why do we need to always update? Can't we tell if the secret needs or doesn't need to be updated?

Reply:
thanks, I have added comparison for StringData and Data fields.